### ISSUE TITLE
Literal expression validation does not match documentation

### ISSUE THREAD

**Issue by dereklieu** [mapbox/mapbox-gl-js#9308]
The style spec documentation for [literal expressions](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#types-literal) says it requires an array or object in the second argument. However, in Studio, an expression like the one below passes validation.

`
[ "literal", "#333333" ]
`

**mapbox-gl-js version**: `@mapbox/mapbox-gl-style-spec@13.11.0`

**browser**: N/A

### Steps to Trigger Behavior

Save this style, upload to Studio, and try clicking on the single layer. 

```
{
    "layers": [
        {
            "id": "this-be-the-problem",
            "layout": {},
            "paint": {
                "fill-extrusion-color": [
                    "case",
                    [
                        "boolean",
                        false
                    ],
                    "#333333",
                    [
                        "literal",
                        "#999999"
                    ]
                ]
            },
            "source": "composite",
            "source-layer": "building",
            "type": "fill-extrusion"
        }
    ],
    "name": "Problematic Literal",
    "sources": {
        "composite": {
            "type": "vector",
            "url": "mapbox://mapbox.mapbox-streets-v8"
        }
    },
    "version": 8
}
```

I'm sorry I don't have a better way to illustrate this, but Studio uses validator functions when you try uploading any style. 

This causes issues in our json editor, which enforces Studio limitations around object literals and only allows arrays in `literal` expressions.

I'm not sure how long this behavior has been around, but it's likely that stricting up this validation could break production styles. It would be worth seeing how many styles this would affect, cc @samanpwbb.

**Comment by asheemmamoowala** [comment]
@dereklieu `literal` expression are most useful when working with literal arrays or objects, but can also be used to represent any other type of Javascript literal. 

Here's a comment from the parsing code:
https://github.com/mapbox/mapbox-gl-js/blob/d5bce5a93eae73566d5e5526149c830e0854667e/src/style-spec/expression/parsing_context.js#L122-L125

Internally, some expressions may be computable at compile time and folded into constants, and continue to be represented by `literal` expressions. This is a legitimate use case. 

I think the documentation is correct to recommend using `literal` for only the cases where it is appropriate. Could this be solved by just updating the documentation?

cc @ryanhamley

**Comment by dereklieu** [comment]
Sounds good - this seems like a docs issue then. We will loosen the validation causing this error in Studio.

**Comment by ryanhamley** [comment]
I'm confused by the Studio validation on this. This is much stricter than how the `literal` expression works.

<img width="571" alt="Screen Shot 2020-02-21 at 11 06 17 AM" src="https://user-images.githubusercontent.com/4523080/75063674-376fec00-549a-11ea-9057-c64ea9cddd02.png">

The expression will accept any `value` type

https://github.com/mapbox/mapbox-gl-js/blob/517b39fbed3981cc06914c9318b384c80f7447f9/src/style-spec/expression/definitions/literal.js#L21-L44

This function shows what is considered a valid `value`:

https://github.com/mapbox/mapbox-gl-js/blob/517b39fbed3981cc06914c9318b384c80f7447f9/src/style-spec/expression/values.js#L34-L68

We might want to adjust the documentation to make it clearer that the `literal` expression is valid with all types, but it seems like the Studio validation may be too strict as well.

**Comment by dereklieu** [comment]
Yeah @ryanhamley my guess based on the code is that we followed the [docs](https://docs.mapbox.com/mapbox-gl-js/style-spec/expressions/#types-literal) which gives the strong impression that you're only supposed to use arrays and objects. Studio doesn't support object types yet, hence the validation.
